comdb2/_ccdb2.pyx

```diff
@@ -153,6 +160,54 @@ cdef class _ParameterValue(object):
                 self.data = PyMem_Malloc(self.size)
                 _bind_datetime(obj, <lib.cdb2_client_datetime_t*>self.data)
                 return
+            elif isinstance(obj, (list, tuple)):
+                self.list_size = len(obj)
+                if 0 == self.list_size:
+                    raise ValueError(f"empty {type(obj).__name__}s cannot be bound")
+
+                if all(isinstance(ele, six.integer_types) for ele in obj):
+                    self.type = lib.CDB2_INTEGER
+                    self.size = sizeof(long long)
+                    self.owner = None
+                    self.data = PyMem_Malloc(self.list_size * self.size)
+                    for l_index in range(self.list_size):
+                        (<long long*>self.data)[l_index] = obj[l_index]
+                    return
+                elif all(isinstance(ele, float) for ele in obj):
+                    self.type = lib.CDB2_REAL
+                    self.size = sizeof(double)
+                    self.owner = None
+                    self.data = PyMem_Malloc(self.list_size * self.size)
+                    for l_index in range(self.list_size):
+                        (<double*>self.data)[l_index] = obj[l_index]
+                    return
+                elif all(isinstance(ele, bytes) for ele in obj):
+                    self.type = lib.CDB2_BLOB
+                    self.size = sizeof(blob_descriptor)
+                    self.owner = None
+                    self.data = PyMem_Malloc(self.list_size * self.size)
+                    for l_index in range(self.list_size):
+                        (<blob_descriptor*>self.data)[l_index].size = len(obj[l_index])
+                        (<blob_descriptor*>self.data)[l_index].data = obj[l_index]
+                    return
+                elif all(isinstance(ele, unicode) for ele in obj):
+                    self.type = lib.CDB2_CSTRING
+                    self.size = sizeof(char*)
+                    # Strings need to be converted to bytes
+                    self.owner = [x.encode('utf-8') for x in obj]
+                    self.data = PyMem_Malloc(self.list_size * self.size)
```

Comment:
What about here, why call `PyMem_Malloc`? I know that the `__dealloc__` was changed so that it calls `PyMem_Free` every time that `self.list_size != -1`, so there is no leak, but aren't we using twice as much memory as we need to once again?

In this case I think `self.owner` is initialized with a list that is a new copy with all of the data we need there, in a Python object that has its reference count set properly, so we can just make `self.data` point to the data held by `self.owner`. Am I missing something?

Reply:
So the python object holds a list of strings. This is allocated on the heap by python during runtime.
For example a = ['apple', 'banana', 'orange']

The pyMem_Malloc creates a C array of pointers. The pointers themselves do not store the entire strings but just the starting memory locations. Each pointer is only 64 bits long even though the underlying string it points to could be 
as long as python wants. So, we aren't allocating memory "twice".
We need to pass an array of pointers or an array of "char*" so that the c api knows where each string starts from.
Then it just reads from the memory allocated in the python list.

How does it know when to stop? This is why we need to do x.encode('utf-8'). This ensures the string terminates
with a null character. something that is common in C-style strings.

For ints and floats we already know they are of fixed length so allocate memory directly for longs, or ints.
Same cannot be said about a strings or bytes hence the usage of pointers.

Basically a char** is an array of char* where each char* is the first byte for each string in our python array.